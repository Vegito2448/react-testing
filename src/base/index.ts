export * from './async-await';
export * from './des-arr';
export * from './functions';
export * from './imp-exp';
export * from './promises';
export * from './template-string';
