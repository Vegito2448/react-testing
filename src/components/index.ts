export * from './MuiMode';
export * from './application';
export * from './counter';
export * from './greet';
export * from './skills';
