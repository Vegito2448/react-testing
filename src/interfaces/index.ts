export * from './skills.types';
export * from './todos.types';
export * from './useCounter.types';
export * from './users.types';
