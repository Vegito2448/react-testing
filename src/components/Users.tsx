import { useEffect, useState } from 'react';
import { User } from '../interfaces';

export const Users = () => {
  const [users, setUsers] = useState<string[]>([]);
  const [error, setError] = useState<string | null>(null);

  useEffect(() => {
    fetch('https://jsonplaceholder.typicode.com/users')
      .then((res): Promise<User[]> => res.json())
      .then((data) => setUsers(data.map((user) => user.name)))
      .catch(() => setError('Error fetching users'));
  }, []);

  return (
    <div>
      <h1>Users</h1>
      {error && <p>{error}</p>}
      <ul>
        {users.map((user) => (
          <li key={user}>{user}</li>
        ))}
      </ul>
    </div>
  );
};
