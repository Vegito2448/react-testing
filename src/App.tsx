import { useEffect, useState } from 'react';
import './App.css';
import { Application, Counter, MuiMode, Skills } from './components';
import { Todo, Todos } from './interfaces';

function App() {
  const [todos, setTodos] = useState<Todo[]>([]);

  useEffect(() => {
    fetch('https://dummyjson.com/todos')
      .then((res) => res.json())
      .then((data: Todos) => {
        setTodos(data.todos);
      });
  }, []);

  return (
    <>
      <MuiMode />
      <h1>Vite + React</h1>
      <Counter />
      <hr />
      {Boolean(todos.length) && <div>Todo List: {todos.length}</div>}
      <hr />
      <Application />
      <hr />
      <Skills title="Skills" skills={['JS', 'ML', 'Py', 'PHP']} />
    </>
  );
}

export default App;
