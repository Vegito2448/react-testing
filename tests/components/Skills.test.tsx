import { logRoles, render, screen } from "@testing-library/react";
import { Skills } from "../../src/components";

describe('Skills', () => {
  const skills = ["JavaScript", "TypeScript", "React"];
  const title = "Skills";
  test('renders correctly', () => {
    render(<Skills title={title} skills={skills} />);
    const titleElement = screen.getByRole('heading', { name: /skills/i });
    expect(titleElement).toBeInTheDocument();
    const listElement = screen.getByRole('list');
    expect(listElement).toBeInTheDocument();
  });

  test('render a list of skills', () => {
    render(<Skills title={title} skills={skills} />);
    const listItemElement = screen.getAllByRole('listitem');
    expect(listItemElement).toHaveLength(skills.length);
    expect(listItemElement[0]).toHaveTextContent('JavaScript');
    expect(listItemElement[1]).toHaveTextContent('TypeScript');
    expect(listItemElement[2]).toHaveTextContent('React');
  });

  test('renders correctly without skills', () => {
    render(<Skills title={title} skills={[]} />);
    const titleElement = screen.getByRole('heading', { name: /skills/i });
    expect(titleElement).toBeInTheDocument();
  });

  test('renders correctly without title', () => {
    render(<Skills skills={skills} />);
    const titleElement = screen.queryByRole('heading', { name: /skills/i });
    expect(titleElement).not.toBeInTheDocument();
    const listElement = screen.getByRole('list');
    expect(listElement).toBeInTheDocument();
  });

  test('renders login button', () => {
    render(<Skills title={title} skills={skills} />);
    const loginButton = screen.getByRole('button', { name: /login/i });
    expect(loginButton).toBeInTheDocument();
  });

  test('renders logout button', async () => {
    render(<Skills title={title} skills={skills} />);
    const loginButton = screen.queryByRole('button', { name: /login/i });
    if (!loginButton) return;
    await loginButton.click();
    const logoutButton = screen.getByRole('button', { name: /logout/i });
    expect(logoutButton).toBeInTheDocument();
    const loginButtonPostClick = screen.queryByRole('button', { name: /login/i });
    expect(loginButtonPostClick).not.toBeInTheDocument();
  });

  test('logout button eventually displayed', async () => {
    const view = render(<Skills title={title} skills={skills} />);
    logRoles(view.container);
    const logoutButton = await screen.findByRole('button', { name: "Logout" }, {
      timeout: 2000
    });
    // screen.debug();
    expect(logoutButton).toBeInTheDocument();
  });

});