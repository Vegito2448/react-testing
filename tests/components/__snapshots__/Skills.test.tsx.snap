// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Skills > logout button eventually displayed 1`] = `
<button>
  Logout
</button>
`;
